johnzon-jsonb/src/main/java/org/apache/johnzon/jaxrs/jsonb/jaxrs/JsonbJaxrsProvider.java

```diff
@@ -187,7 +189,23 @@ public long getSize(final T t, final Class<?> type, final Type genericType, fina
 
     @Override
     public T readFrom(final Class<T> type, final Type genericType, final Annotation[] annotations, final MediaType mediaType,
-            final MultivaluedMap<String, String> httpHeaders, final InputStream entityStream) throws IOException, WebApplicationException {
+            final MultivaluedMap<String, String> httpHeaders, InputStream entityStream) throws IOException, WebApplicationException {
+        if (entityStream.markSupported()) {
```

Comment:
hey @mkarg, I would like to mention these few points before moving forward, please let me know what you think about it:

1. here the stream will rarely be markable (with cxf it will not most of the time) so we can surely skip this case
2. we already have a pushback inputstream for the jsonparser so i would consider adding a callback here we can register through the parser map. High level it would be a callback propagated through https://github.com/apache/johnzon/blob/master/johnzon-jsonb/src/main/java/org/apache/johnzon/jsonb/JohnzonBuilder.java#L837 from Jsonb default instance to the parser and it would be called https://github.com/apache/johnzon/blob/156c7af77322fc0d11819756fc48fb18bd617b71/johnzon-core/src/main/java/org/apache/johnzon/core/RFC4627AwareInputStreamReader.java#L51 (this stream reader is created by the parser). The callback can then throw an exception the readFrom can catch and process in an exceptional way (see point 3)
3. trivial impl of this exceptional way can be throw new NoContentException but it also makes sense to have a way to respect the quote from the spec you did in the ticket and return an "empty" object IMHO, I assume it must be an option of the provider and handled in the catch of point 2 as well
4. you forget to commit the associated test? ;)

wdyt?

Reply:
@rmannibucau Thank you for the quick review. :-)
1. Actually I tried it out with the following result: Payara 4 really provides a markable stream, pure Jersey does not (or vice versa, I haven't actually written down), so dropping the mark case would _actually_ turn down performance at least for *some* configurations. So I would vote for keeping this case to provide *best possible* performance in each particular enviroment.
2. I will look into that, but quickly answered this sounds complex, error-prone and slow (try-catch IMHO is slower than double buffering). What would be the _actual_ benefit compared to my original proposal?
3. The intention of *this* PR is to provide JAX-RS 2.1 compliance **ASAP**. I see your point here and would love to have a more sophisticated solution, which e. g. checks if the requested class would allow creation of empty instances or not. But as the currently wrong code does not have this solution, and as it sounds non-trivial and slow, I think we must spend rather long time to get that attempt right _and_ fast _and_ bug-free. So IMHO we should do that in a _later_ PR _ontop_ of this one.
4. What is a test? ;-)

Reply:
1. can be solved by a bench but from what you say mainstream is not markable (only payara?) so it shouldn't be the first citizen probably?
2. the benefit is the speed, overhead in O(1) vs O(size of the stream) (you can bench it on small, medium, huge streams)
3. this is not that hard or slow ;). My main point is we can't slow down a stream more than we do already, it is the slowest part we have and here we try to handle a case which never happens (literally) so any overhead in O(n) is negative for real users.
4. I didnt read that answer :D

Reply:
Thanks for the explanation.

In fact I do not see why the upfront markable check actually bothers you: It eats up rather zero time as the implementation in the usual suspects would be `return false;`? Anyways, I have no strong feeling about removing that case if you hate it so much. ;-)

Before I deep-dive into that code (I will do definitively): So what you claim is that `PushbackInputBuffer` actually is O(size of _stream_) -which would be a desaster!- instead of O(size of buffer) -which would be negligible as the buffer size is 1 byte-? Really? Didn't know that! Actually I should open a PR against the JRE to change that! ;-)

Reply:
The markable point is just in terms of maintenance but no other blocker so you can keep it, just ensure it works when false ;)

The pushback overhead is indeed o(n) for the same reason an arraylist if often replaced by an array in optimized impl (checkopened+length check), we dont need it here so it is just wasting resources to stack them.

N is the stream size, not the buffer one ;).

Hope it is clearer.

Reply:
I checked the code and you are right, this is something that we should address to prevent the O(n) overhead. I will look into that definitively, but to not further delay fixing of this bug (my application expects to catch `NoContentException` so I'm stuck without this) I'd kindly request that I may fix the performance regression *later this week*.